README.md
# CRC_t - is C++ template for calculation CRC any sizes 1-64 bits


## Description

`CRC_t` is C++ template for calculation CRC any sizes(width) 1-64 bits.

#### Features of the implementation:

 - The code has no dependencies, and imprisoned in the one file.
 - The following implementations are supported:
    * **CRCImplBits** - loop for 8 bits in byte (no table)
    * **CRCImplTable4** - table for half byte (16 elements)
    * **CRCImplTable8** - std table for byte (256 elements)
 - All the parameters of CRC passed as template parameters. This allows the compiler to make a very strong optimization.
 Example ModBus-CRC for architecture ARMv7 (GCC 4.9.2) calculate CRC for a buffer(array) will loop from 9 commands in assembler.


#### Limitations of the implementation:

- After the parameterization of the template, you cannot change the parameters of the CRC.


If you need a class to work with different CRC algorithms, which can change the algorithm parameters in the dynamics (Run-Time) see:
[uCRC_t](https://github.com/KoynovStas/uCRC_t)


Template parameters is the standard Specifications algorithms CRC as described in Ross N. Williams [A PAINLESS GUIDE TO CRC ERROR DETECTION ALGORITHMS](./doc/crc_v3.txt)


**Ross N. Williams:**

> By putting all these pieces together we end up with the parameters of the algorithm:
>
>   - **NAME:** This is a name given to the algorithm. A string value.
>
>   - **WIDTH:** This is the width of the algorithm expressed in bits.
>   This is one less than the width of the Poly.
>
>   - **POLY:** This parameter is the poly. This is a binary value that
>   should be specified as a hexadecimal number. The top bit of the
>   poly should be omitted. For example, if the poly is 10110, you
>   should specify 06. An important aspect of this parameter is that it
>   represents the unreflected poly; the bottom bit of this parameter
>   is always the LSB of the divisor during the division regardless of
>   whether the algorithm being modelled is reflected.
>
>   - **INIT:** This parameter specifies the initial value of the register
>   when the algorithm starts. This is the value that is to be assigned
>   to the register in the direct table algorithm. In the table
>   algorithm, we may think of the register always commencing with the
>   value zero, and this value being XORed into the register after the
>   N'th bit iteration. This parameter should be specified as a
>   hexadecimal number.
>
>   - **REFIN:** This is a boolean parameter. If it is FALSE, input bytes are
>   processed with bit 7 being treated as the most significant bit
>   (MSB) and bit 0 being treated as the least significant bit. If this
>   parameter is FALSE, each byte is reflected before being processed.
>
>   - **REFOUT:** This is a boolean parameter. If it is set to FALSE, the
>   final value in the register is fed into the XOROUT stage directly,
>   otherwise, if this parameter is TRUE, the final register value is
>   reflected first.
>
>   - **XOROUT:** This is an W-bit value that should be specified as a
>   hexadecimal number. It is XORed to the final register value (after
>   the REFOUT) stage before the value is returned as the official
>   checksum.
>
>   - **CHECK:** This field is not strictly part of the definition, and, in
>   the event of an inconsistency between this field and the other
>   field, the other fields take precedence. This field is a check
>   value that can be used as a weak validator of implementations of
>   the algorithm. The field contains the checksum obtained when the
>   ASCII string "123456789" is fed through the specified algorithm
>   (i.e. 313233... (hexadecimal)).



**Template CRC_t has the following parameters:**
```C++
template <Bits, Poly, Init, RefIn, RefOut, XorOut, Impl = CRCImplTable8>
```


**The class has the following public methods:**
```C++
// get param CRC
static constexpr uint8_t  get_bits()    noexcept { return Bits;  }
static constexpr CRC_Type get_poly()    noexcept { return Poly;  }
static constexpr CRC_Type get_init()    noexcept { return Init;  }
static constexpr CRC_Type get_xor_out() noexcept { return XorOut;}
static constexpr bool     get_ref_in()  noexcept { return RefIn; }
static constexpr bool     get_ref_out() noexcept { return RefOut;}

static constexpr CRC_Type get_top_bit() noexcept { return (CRC_Type)1 << (Bits - 1);         }
static constexpr CRC_Type get_crc_mask()noexcept { return ( (get_top_bit() - 1) << 1) | 1;   }
static constexpr CRC_Type get_crc_init()noexcept { return RefIn ? reflect(Init, Bits) : Init;}

constexpr CRC_Type get_check() const noexcept;//crc for ASCII string "123456789" (3132..39(in hex))

// Calculate methods
constexpr CRC_Type get_crc(const void* data, size_t len) const noexcept;
int                get_crc(CRC_Type &crc, const char* file_name) const noexcept;
int                get_crc(CRC_Type &crc, const char* file_name, void* buf, size_t size_buf) const noexcept;

// Calculate for chunks of data
constexpr        CRC_Type get_raw_crc(const void* data, size_t len) const noexcept;
constexpr        CRC_Type get_raw_crc(const void* data, size_t len, CRC_Type raw_crc) const noexcept;
static constexpr CRC_Type get_end_crc(CRC_Type raw_crc) noexcept;
```

More details see: **[crc_t.h](./crc_t.h)**


## Usage

**To start working, perform the following steps:**

1. You must select the appropriate version:
  + C++98 version - 2.x
  + C++11 version - 3.x
  + C++14 version - 4.x
    > Where .x maximum minor version at the moment
2. You need to include **[crc_t.h](./crc_t.h)** file in your **.cpp** file
3. Parameterize a template (see an examples)

**Note:**
> The parameters for the template correspond to the `Ross N. Williams` specification.
> A list of CRC algorithms for this specification can be found here: [Catalogue CRC algorithms](http://reveng.sourceforge.net/crc-catalogue/all.htm)
> or see [crc_list.h](./tests/crc_list.h)


Since from version 2.0 you can choose a different calculation algorithm.
The default implementation parameter is `Impl = CRCImplTable8`.
You can select the following algorithm:
* **CRCImplBits** - loop for 8 bits in byte (no table)
* **CRCImplTable4** - table for half byte(4-bits) (16 elements)
* **CRCImplTable8** - std table for byte (8-bits) (256 elements)

Result of speed test: [benchmark.md](./tests/benchmark.md)


## Examples

**Get CRC32 for file:**

```C++
uint32_t crc;

CRC_t<32, 0x04C11DB7, 0xFFFFFFFF, true, true, 0xFFFFFFFF> ucrc;

int err = ucrc.get_crc(crc, "std_file_to_test_crc");

if( !err )
    //uses crc
```


**Note: methods for calculate CRC file**

>  ```C++
>  int get_crc(CRC_Type &crc, const char* file_name) const noexcept
>  ```
>  These method are reentrant. He use a buffer on the stack.
>  The buffer size is 4 Kib (4096 bytes) which is optimal for most systems.
>  If you have a buffer or needs aligned buffer, you can use the following method:
>
>  ```C++
>  int get_crc(CRC_Type &crc, const char* file_name, void* buf, size_t size_buf) const noexcept
>  ```


**Get CRC32 for single buf:**

```C++
char buf[len_of_buf]; //bla bla

uint32_t crc;

CRC_t<32, 0x04C11DB7, 0xFFFFFFFF, true, true, 0xFFFFFFFF> ucrc;

crc = ucrc.get_crc(buf, len_of_buf);
//uses crc
```


**Get CRC-ModBus for single buf:**

```C++
char buf[len_of_buf]; //bla bla

uint16_t crc;

CRC_t<16, 0x8005, 0xffff, true, true, 0x0> ucrc;

crc = ucrc.get_crc(buf, len_of_buf);
//uses crc
```



**Get CRC-8 for buf(s) (chunks):**

**Note:**
>  when the method is used `CRC_Type get_raw_crc(const void* data, size_t len, CRC_Type raw_crc)`
>  for the first byte (or chunk of data) **crc** param must be obtained through a method `get_crc_init()`
>  and in the end you need to call the method: `get_end_crc()`:

```C++
char buf[len_of_buf];   //bla bla
char buf2[len_of_buf2]; //bla bla

uint8_t crc;

CRC_t<8, 0x7, 0x0, false, false, 0x0> ucrc;

crc = ucrc.get_crc_init();
crc = ucrc.get_raw_crc(buf,  len_of_buf,  crc);  //first chunk
crc = ucrc.get_raw_crc(buf2, len_of_buf2, crc);  //second chunk
crc = ucrc.get_end_crc(crc);
//uses crc

//Since from version 2.0
crc = ucrc.get_raw_crc(buf,  len_of_buf);        //first chunk
crc = ucrc.get_raw_crc(buf2, len_of_buf2, crc);  //second chunk
crc = ucrc.get_end_crc(crc);
//uses crc
```


**Note for bare-metal:**

Since version 4.0, table implementations have two constructors.
The default constructor is `constexpr` (table calculation in compile-time),
the second one is needed to call the constructor in run-time(table calculation in run-time).
This allows you to get various options for working, for example, place a table in the DTCM-RAM by specifying the desired section.

Example for gcc:

```C++
using CRC32_t = CRC_t<32, 0x4C11DB7, 0xFFFFFFFF, false, false, 0x0>;
#define DTCM_RAM  __attribute__((section(".DTCM_data")))

const CRC32_t          sw_crc32{}; //table in section .text      (Flash)    (table calculation in compile-time)
const CRC32_t DTCM_RAM sw_crc32{}; //table in section .DTCM_data (DTCM-RAM) (table calculation in compile-time)
const CRC32_t DTCM_RAM sw_crc32{1};//table in section .DTCM_data (DTCM-RAM) (table calculation in run-time)
const CRC32_t          sw_crc32{1};//table in section .bss       (RAM)      (table calculation in run-time)
```


More details can be found in the test application: **[crc_test.cpp](./tests/crc_test.cpp)**



## Tests

To run tests, perform the following steps:

```
cd tests
cmake . -B ./build
cmake --build build
```

On Windows for MinGW you can generate `MinGW Makefiles`:
```
cmake . -B ./build -G "MinGW Makefiles"
```



## License

[BSD-3-Clause](./LICENSE).


CHANGELOG.md
# Change Log


[**CRC_t**](https://github.com/KoynovStas/CRC_t) - is C++ template for calculation CRC any sizes(width) 1-64 bits.

License: BSD 3-Clause

Copyright (c) 2015, Koynov Stas - skojnov@yandex.ru



---
## [v4.0](https://github.com/KoynovStas/CRC_t/tree/v4.0) (2023-05-27)
[Full Changelog](https://github.com/KoynovStas/CRC_t/compare/v3.0...v4.0)


### New
 - New: Since version 4.0 required `C++14`
 - New: methods `get_crc` is `constexpr`
 - New: method `reflect` is full `constexpr`
 - New: add `constexpr` for method `get_end_crc`
 - New: add `constexpr` ctor for table `Impl` (see note for bare-metall in README)

Since ver 4.0 methods `get_crc` is `constexpr` This allows you to calculate CRC for `constexpr` variables in compile time.
This works in gcc 10. But it doesn't work in gcc 11. (reason for prohibiting use of `static_cast` from `const void*` in `constexpr` functions)
But this feature should appear in C++26 see proposal: [P2738](https://github.com/cplusplus/papers/issues/1431)




## [v3.0](https://github.com/KoynovStas/CRC_t/tree/v3.0) (2023-05-23)
[Full Changelog](https://github.com/KoynovStas/CRC_t/compare/v2.1...v3.0)


### New
 - New: Since version 3.0 required `C++11`
 - New: add `clean_up` for ASan
 - New: add `constexpr` variant of `reflect` method (only for constants)
 - Fix: deleted `crc_init` member (method `get_crc_init()` is full `constexpr`


### Refactoring
 - use std `static_assert`
 - replace `NULL` -> `nullptr`




## [v2.1](https://github.com/KoynovStas/CRC_t/tree/v2.1) (2023-05-12)
[Full Changelog](https://github.com/KoynovStas/CRC_t/compare/v2.0...v2.1)


### New
 - New: use cmake for build tests
 - Fix: use `__INCLUDE_LEVEL__` in `STATIC_ASSERT` for fix ODR in headers


### Refactoring
 - replaced `std::string` to `const char*` in Proxy
 - use `stest_printf` for embedded (was be `std::cerr`)
 - deleted `get_cnt_impl()`, add const `CNT_IMPL`
 - deleted `sstream`, use `TEST_ASSERTF` with `printf`
 - add `static` prefix for some methods


### Tests
 - add to CRC list `CRC-64/MS`
 - add to CRC list `CRC-64/REDIS`
 - add to CRC list `CRC-32/MEF`
 - add to CRC list `CRC-16/M17`
 - add to CRC list `CRC-8/HITAG`




## [v2.0](https://github.com/KoynovStas/CRC_t/tree/v2.0) (2020-05-01)
[Full Changelog](https://github.com/KoynovStas/CRC_t/compare/v1.2...v2.0)


### New
- New: Started using pattern [CRTP](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) this allows the use of static polymorphism. And this allows you to extract each algorithm to a separate class!
- New: extracted base class `CRCBase_t` from `CRC_t`
- New: add template parameter `Impl` (Implementation of algorithm)
- New: add implementation of algorithm **CRCImplTable4** - table for half byte (16 elements)
- New: add implementation of algorithm **CRCImplBits** - loop for 8 bits in byte (no table)

- The current version are supported:
   * **CRCImplBits** - loop for 8 bits in byte (no table) - **new**
   * **CRCImplTable4** - table for half byte (16 elements) - **new**
   * **CRCImplTable8** - std table for byte (256 elements)


### Refactoring
- Ref: join `init_xxx_crc_table` to one method (small code)
- Ref: moved `get_raw_normal`/`reflect_crc` from `ImplBits` to Base class


### Tests
 - add to tests **CRCImplBits** strategy
 - add to tests **CRCImplTable4** strategy
 - add `test_crc_impl_1byte`
 - add `test_crc_impl_data_xyz_256`
 - add `test_crc_impl_data_xxx_256`
 - add `test_crc_for_cunks2` for test wrapper `get_raw_crc`
 - add `test_crc_speed` - benchmark for calc speed for 1GB data




## [v1.2](https://github.com/KoynovStas/CRC_t/tree/v1.2) (2018-01-11)
[Full Changelog](https://github.com/KoynovStas/CRC_t/compare/v1.1...v1.2)



### New
- deleted `name` field - user self can add it (if need)


### BugFix
 - BugFix: bad error message if test is fail (`ss` was be on stack)


### Refactoring
 - `reflect` method
 - use `std::ifstream` for calculate crc for file
 - delete prefix `register` - old style
 - moved var `shift` to `get_raw_crc()` method
 - moved calculate `crc_mask` and `top_bit` to methods
 - divided the `init_crc_table` function into **two** functions `normal\reflected`
 - rename `get_final_crc` -> `get_end_crc`


### Tests
 - add to CRC list `CRC-8/BLUETOOTH`
 - add to CRC list `CRC-8/NRSC-5`
 - add to CRC list `CRC-16/NRSC-5`
 - add to CRC list `CRC-17/CAN-FD`
 - add to CRC list `CRC-21/CAN-FD`
 - add to CRC list `CRC-24/OS-9`




## [v1.1](https://github.com/KoynovStas/CRC_t/tree/v1.1) (2017-05-03)
[Full Changelog](https://github.com/KoynovStas/CRC_t/compare/v1.0...v1.1)


### Added
- add method `get_check` - crc for ASCII string '123456789'


### Refactoring
 - Replacement `(const char *)` to `(const void *)` - for ease use methods: `get_crc` and `get_raw_crc`

crc_t.h
/*
 * CRC_t is C++ template for calculation CRC any sizes(width) 1-64 bits.
 *
 *
 * version 4.0
 *
 *
 * BSD 3-Clause License
 *
 * Copyright (c) 2015, Koynov Stas - skojnov@yandex.ru
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#ifndef CRC_T_H
#define CRC_T_H

#include <cstdint>
#include <cerrno>
#include <array>
#include <fstream>    // for std::ifstream
#include <ios>        // for std::ios_base, etc.





template<uint8_t Bits_minus_1_div_8> struct CRC_Type_helper{ using value_type = uint64_t; }; // default

template<> struct CRC_Type_helper<0> { using value_type = uint8_t;  }; //for Bits 1..8
template<> struct CRC_Type_helper<1> { using value_type = uint16_t; }; //for Bits 9..16
template<> struct CRC_Type_helper<2> { using value_type = uint32_t; }; //for Bits 17..24
template<> struct CRC_Type_helper<3> { using value_type = uint32_t; }; //for Bits 25..32


#define CRC_TYPE typename CRC_Type_helper< (Bits-1)/8 >::value_type





/*
 * CRCBase_t - is the base class that we will use in various algorithms.
 * This class accepts an Impl template parameter (use Curiously Recurring Template Pattern (CRTP)).
 * This allows us to use the get_raw_crc_impl() method from the implementation class without using virtual methods.
 */
template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
class CRCBase_t
{
    static_assert((Bits >= 1) && (Bits <= 64), "Bits must be in the range [1,64]");


    public:
        using CRC_Type = CRC_TYPE;

        // get param CRC
        static constexpr uint8_t  get_bits()    noexcept { return Bits;  }
        static constexpr CRC_Type get_poly()    noexcept { return Poly;  }
        static constexpr CRC_Type get_init()    noexcept { return Init;  }
        static constexpr CRC_Type get_xor_out() noexcept { return XorOut;}
        static constexpr bool     get_ref_in()  noexcept { return RefIn; }
        static constexpr bool     get_ref_out() noexcept { return RefOut;}

        static constexpr CRC_Type get_top_bit() noexcept { return (CRC_Type)1 << (Bits - 1);         }
        static constexpr CRC_Type get_crc_mask()noexcept { return ( (get_top_bit() - 1) << 1) | 1;   }
        static constexpr CRC_Type get_crc_init()noexcept { return RefIn ? reflect(Init, Bits) : Init;}

        constexpr CRC_Type get_check() const noexcept;//crc for ASCII string "123456789" (3132..39(in hex))

        // Calculate methods
        constexpr CRC_Type get_crc(const void* data, size_t len) const noexcept;
        int                get_crc(CRC_Type &crc, const char* file_name) const noexcept;
        int                get_crc(CRC_Type &crc, const char* file_name, void* buf, size_t size_buf) const noexcept;

        // Calculate for chunks of data
        constexpr        CRC_Type get_raw_crc(const void* data, size_t len) const noexcept;
        constexpr        CRC_Type get_raw_crc(const void* data, size_t len, CRC_Type raw_crc) const noexcept;
        static constexpr CRC_Type get_end_crc(CRC_Type raw_crc) noexcept;


    protected:
        static constexpr CRC_Type reflect(CRC_Type data, uint8_t num_bits) noexcept;
        static constexpr CRC_Type get_raw_normal_crc   (uint8_t byte)      noexcept;
        static constexpr CRC_Type get_raw_reflected_crc(uint8_t byte)      noexcept;


    private:
        int get_crc(CRC_Type &crc, std::ifstream& ifs, void* buf, size_t size_buf) const noexcept;
};





template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
constexpr CRC_TYPE CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::get_check() const noexcept
{
    constexpr uint8_t data[] = {0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39};

    return get_crc(data, sizeof(data));
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
constexpr CRC_TYPE CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::
get_crc(const void* data, size_t len) const noexcept
{
    CRC_Type crc = get_raw_crc(data, len, get_crc_init());

    return get_end_crc(crc);
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
int CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::
get_crc(CRC_Type &crc, const char *file_name) const noexcept
{
    char buf[4096];

    return get_crc(crc, file_name, buf, sizeof(buf));
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
int CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::
get_crc(CRC_Type &crc, const char* file_name, void* buf, size_t size_buf) const noexcept
{
    std::ifstream ifs(file_name, std::ios_base::binary);

    if( !ifs || !buf || !size_buf)
    {
        errno = EINVAL;
        return -1;
    }

    return get_crc(crc, ifs, buf, size_buf);
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
int CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::
get_crc(CRC_Type &crc, std::ifstream& ifs, void* buf, size_t size_buf) const noexcept
{
    crc = get_crc_init();

    while( ifs )
    {
        ifs.read(static_cast<char *>(buf), size_buf);
        crc = get_raw_crc(buf, ifs.gcount(), crc);
    }

    crc = get_end_crc(crc);

    return (ifs.rdstate() & std::ios_base::badbit);  //return 0 if all good
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
constexpr CRC_TYPE CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::
get_raw_crc(const void* data, size_t len) const noexcept
{
    return get_raw_crc(data, len, get_crc_init());
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
constexpr CRC_TYPE CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::
get_raw_crc(const void* data, size_t len, CRC_Type raw_crc) const noexcept
{
    //use Curiously Recurring Template Pattern (CRTP)
    return static_cast<const Impl*>(this)->get_raw_crc_impl(data, len, raw_crc);
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
constexpr CRC_TYPE CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::
get_end_crc(CRC_Type raw_crc) noexcept
{
    if(RefOut^RefIn)
        raw_crc = reflect(raw_crc, Bits);

    raw_crc ^= XorOut;
    raw_crc &= get_crc_mask(); //for CRC not a multiple of a byte (8 bits)

    return raw_crc;
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
constexpr CRC_TYPE CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::
reflect(CRC_Type data, uint8_t num_bits) noexcept
{
    CRC_Type reflection = 0;

    while( num_bits-- )
    {
        reflection = (reflection << 1) | (data & 1);
        data >>= 1;
    }

    return reflection;
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
constexpr CRC_TYPE CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::
get_raw_normal_crc(uint8_t byte) noexcept
{
    CRC_Type crc = 0;

    for(int bit = 0x80; bit; bit >>= 1)
    {
        if( byte & bit )
            crc ^= get_top_bit();

        if( crc & get_top_bit() )
        {
            crc <<= 1;
            crc ^= Poly;
        }
        else
            crc <<= 1;
    }

    return crc & get_crc_mask(); //for CRC not a multiple of a byte (8 bits);
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut, class Impl>
constexpr CRC_TYPE CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut, Impl>::
get_raw_reflected_crc(uint8_t byte) noexcept
{
    constexpr CRC_Type ref_poly = reflect(Poly, Bits);

    CRC_Type crc = byte;

    for(int bit = 0x80; bit; bit >>= 1)
    {
        if( crc & 1 )
        {
            crc >>= 1;
            crc ^= ref_poly;
        }
        else
            crc >>= 1;
    }

    return crc;
}





template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut>
class CRCImplTable8: public CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut,
                                      CRCImplTable8<Bits, Poly, Init, RefIn, RefOut, XorOut> >
{
    public:
        using CRC_Type = CRC_TYPE;

        constexpr CRCImplTable8() noexcept : crc_table{} { init_crc_table(); }

        //Ctor to init table in RunTime (for bare-metal or embedded)
        explicit CRCImplTable8(int) { init_crc_table(); }

        constexpr CRC_Type get_raw_crc_impl(const void* data, size_t len, CRC_Type crc) const noexcept;


    private:
        CRC_Type crc_table[256];

        constexpr void init_crc_table() noexcept;
};




template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut>
constexpr CRC_TYPE CRCImplTable8<Bits, Poly, Init, RefIn, RefOut, XorOut>::
get_raw_crc_impl(const void* data, size_t len, CRC_Type crc) const noexcept
{
    auto buf = static_cast< const uint8_t* >(data);

    constexpr int shift = (Bits > 8) ? (Bits - 8) : (8 - Bits);


    if(Bits > 8)
    {
        if(RefIn)
            while (len--)
                crc = (crc >> 8) ^ crc_table[(crc ^ *buf++) & 0xff];
        else
            while (len--)
                crc = (crc << 8) ^ crc_table[((crc >> shift) ^ *buf++) & 0xff];
    }
    else
    {
        if(RefIn)
            while (len--)
                crc = crc_table[crc ^ *buf++];
        else
            while (len--)
                crc = crc_table[(crc << shift) ^ *buf++];
    }


    return crc;
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut>
constexpr void CRCImplTable8<Bits, Poly, Init, RefIn, RefOut, XorOut>::init_crc_table() noexcept
{
    //Calculation of the CRC table for byte.
    for(int byte = 0; byte < 256; byte++)
    {
        if(RefIn)
            crc_table[byte] = this->get_raw_reflected_crc(byte); // Reflected CRC table for byte.
        else
            crc_table[byte] = this->get_raw_normal_crc(byte);    // Normal CRC table for byte.
    }
}





template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut>
class CRCImplBits: public CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut,
                                    CRCImplBits<Bits, Poly, Init, RefIn, RefOut, XorOut> >
{
    public:
        using CRC_Type = CRC_TYPE;

        constexpr CRC_Type get_raw_crc_impl(const void* data, size_t len, CRC_Type crc) const noexcept;
};





template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut>
constexpr CRC_TYPE CRCImplBits<Bits, Poly, Init, RefIn, RefOut, XorOut>::
get_raw_crc_impl(const void* data, size_t len, CRC_Type crc) const noexcept
{
    auto buf = static_cast< const uint8_t* >(data);

    constexpr int shift = (Bits > 8) ? (Bits - 8) : (8 - Bits);


    if(Bits > 8)
    {
        if(RefIn)
            while (len--)
                crc = (crc >> 8) ^ this->get_raw_reflected_crc((crc ^ *buf++) & 0xff);
        else
            while (len--)
                crc = (crc << 8) ^ this->get_raw_normal_crc(((crc >> shift) ^ *buf++) & 0xff);
    }
    else
    {
        if(RefIn)
            while (len--)
                crc = this->get_raw_reflected_crc(crc ^ *buf++);
        else
            while (len--)
                crc = this->get_raw_normal_crc((crc << shift) ^ *buf++);
    }

    return crc;
}





template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut>
class CRCImplTable4: public CRCBase_t<Bits, Poly, Init, RefIn, RefOut, XorOut,
                                      CRCImplTable4<Bits, Poly, Init, RefIn, RefOut, XorOut> >
{
    public:
        using CRC_Type = CRC_TYPE;

        constexpr CRCImplTable4() noexcept : crc_table{} { init_crc_table(); }

        //Ctor to init table in RunTime (for bare-metal or embedded)
        explicit CRCImplTable4(int) { init_crc_table(); }

        constexpr CRC_Type get_raw_crc_impl(const void* data, size_t len, CRC_Type crc) const noexcept;


    private:
        CRC_Type crc_table[16];

        constexpr void init_crc_table() noexcept;
};





template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut>
constexpr CRC_TYPE CRCImplTable4<Bits, Poly, Init, RefIn, RefOut, XorOut>::
get_raw_crc_impl(const void* data, size_t len, CRC_Type crc) const noexcept
{
    auto buf = static_cast< const uint8_t* >(data);

    constexpr int shift = (Bits > 4) ? (Bits - 4) : (4 - Bits);


    if(Bits > 4)
    {
        if(RefIn)
            while (len--)
            {
                crc = crc_table[(crc ^ *buf       ) & 0x0f] ^ (crc >> 4);
                crc = crc_table[(crc ^ (*buf >> 4)) & 0x0f] ^ (crc >> 4);
                buf++;
            }
        else
            while (len--)
            {
                crc = crc_table[((crc >> shift) ^ (*buf >> 4)) & 0x0f] ^ (crc << 4);
                crc = crc_table[((crc >> shift) ^ *buf       ) & 0x0f] ^ (crc << 4);
                buf++;
            }
    }
    else//Bits <= 4
    {
        if(RefIn)
            while (len--)
            {
                crc = crc_table[(crc ^ *buf       ) & 0x0f];
                crc = crc_table[(crc ^ (*buf >> 4)) & 0x0f];
                buf++;
            }
        else
            while (len--)
            {
                crc = crc_table[((crc << shift) ^ (*buf >> 4)) & 0x0f];
                crc = crc_table[((crc << shift) ^ *buf       ) & 0x0f];
                buf++;
            }
    }

    return crc;
}



template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut>
constexpr void CRCImplTable4<Bits, Poly, Init, RefIn, RefOut, XorOut>::init_crc_table() noexcept
{
    //Calculation of the CRC table for half byte.
    for(int byte = 0; byte < 16; byte++)
    {
        if(RefIn)
            crc_table[byte] = this->get_raw_reflected_crc(byte*16); //see note1
        else
            crc_table[byte] = this->get_raw_normal_crc(byte);
    }

    //note1: *16 it's equivalent function for calculating CRC for 4 bits (the result of symmetry)
}





template <uint8_t Bits, CRC_TYPE Poly, CRC_TYPE Init, bool RefIn, bool RefOut, CRC_TYPE XorOut,
          template<uint8_t, CRC_TYPE, CRC_TYPE, bool, bool, CRC_TYPE> class Impl = CRCImplTable8>
class CRC_t: public Impl<Bits, Poly, Init, RefIn, RefOut, XorOut>
{
    public:
        using Impl<Bits, Poly, Init, RefIn, RefOut, XorOut>::Impl; //for using explicit ctors!
};





#endif // CRC_T_H
